### ISSUE TITLE
Various asJava return types are unsound

### ISSUE THREAD

**Issue by S11001001** [scala/bug#10553]
Several casts in [`AsJavaConverters`](https://github.com/scala/scala/blob/v2.12.3/src/library/scala/collection/convert/AsJavaConverters.scala) are unsound, as demonstrated by the below, [run on scastie](https://scastie.scala-lang.org/ytHt5LscRqKC3iH2SzLO7g).

```scala
import scala.collection.JavaConverters._

object Test {
  // seqAsJavaList is unsound
  def uc1[A, B]: A => B = a => {
    val aa = new java.util.ArrayList[B]
    (aa.asScala: collection.Seq[Any]).asJava.add(a)
    aa.get(0)
  }

  // mapAsJavaMap is unsound
  def uc2[A, B]: A => B = a => {
    val aa = new java.util.HashMap[Unit, B]
    (aa.asScala: collection.Map[Unit, Any]).asJava.put((), a)
    aa.get(())
  }

  // asJavaCollection is unsound
  def uc3[A, B]: A => B = a => {
    val aa: java.util.Collection[B] = new java.util.ArrayList[B]
    (aa.asScala: collection.Iterable[Any]).asJavaCollection.add(a)
    aa.iterator.next
  }

  val test1 = uc1[(Unit, Int), (Unit, Nothing)](((), 42))
  val test2 = uc2[(Unit, Int), (Unit, Nothing)](((), 42))
  val test3 = uc3[(Unit, Int), (Unit, Nothing)](((), 42))
}
```

The other `asInstanceOf`s (e.g. `asJavaIterator`) can be similarly exploited, but require a (100% checked) cast to do so.

This can be fixed in a few ways.

1. Change the signatures to only admit statically-immutable arguments, e.g. `seqAsJavaList` would only accept `immutable.Seq`s as arguments.  Then the round-trip `case` would be eliminated, because it could never happen.  That would be inconvenient for holders of e.g. `collection.Seq`, though.  It would also be binary-incompatible.
2. Change the signatures to return upper-bounded existentials, e.g. `seqAsJavaList` would return `ju.List[_ <: A]`.  The round-trip `case`s would be preserved, with the casts eliminated.  This is binary-compatible, but still a burden on holders of `immutable.Seq` et al, because this unsoundness doesn't apply to them.
3. Split the signatures, e.g. instead of `seqAsJavaList[A](collection.Seq[A])ju.List[A]`, you'd have `iseqAsJavaList[A](collection.immutable.Seq[A])ju.List[A]` and `mseqAsJavaList[A](collection.mutable.Seq[A])ju.List[A]`.  This can't be exploited, but again is slightly inconvenient for holders of `collection.Seq`.  Some of the `mutable` variants already exist, indeed.
4. A combination of # 2 and # 3, e.g. a higher-priority conversion without the existential if you have a statically `immutable` or `mutable` `Seq`, with the lower-priority one producing `ju.List[_ <: A]`.  This yields the most flexible API for users, but if you don't care so much about supporting `collection.Seq` et al, # 3 by itself might be a useful simplification.

(Tested with 2.12.3.)

**Comment by SethTisue** [comment]
`JavaConverters` is deprecated in Scala 2.13+. I wonder if the same issue is demonstrable using its replacement, `scala.jdk.CollectionConverters`

**Comment by lrytz** [comment]
Yes, it's the same with `jdk.CollectionConverters`. I don't see any `asInstanceOf`s, but the unsoundness is the unchecked type argument here: https://github.com/scala/scala/blob/15538c7365bbb03f37d06b59f050241683972c71/src/library/scala/collection/convert/AsJavaConverters.scala#L147

If we remove that case from the pattern match we end up with a new `SeqWrapper`, which extends `java.util.AbstractList`, which throws an `UnsupportedOperationException` on calling `add`.
